### ISSUE TITLE
Idea: Support for hyperbolic soccer ball and similar tillings

### ISSUE THREAD

**Issue by ChristopherKing42** [dmishin/hyperbolic-ca-simulator#1]
https://en.wikipedia.org/wiki/Truncated_order-7_triangular_tiling#Hyperbolic_soccerball

**Comment by dmishin** [comment]
In truncated tilings, different cells have different number of neighbors, which makes them less interesting for CAs: you would probably need a separate rule for each cell type, so more intrinsic complexity. Are there any signs that truncated tilings could be interesting? I am not sure that the result worth the effort.

The hard part is addressing cells of truncated tiling.

Currently, cell address is a word in the triangle group of a tiling. For tiling {n,m} it is a group, generated by {aa = bb = cc = (ab)^n = (bc)^m = (ac)^2 = 1}

Such word identifies one of 2n triangles forming an n-gonal cell. The shortest of all triangle addresses serves as a cell address.

Address of a truncated tiling cell can be formed by extending original cell address with an additional flag, telling what kind of cells (red or yellow) is being addressed.

**Comment by ChristopherKing42** [comment]
Ah, I guess you're right. I was just thinking it would be more (but not fully) Euclidean.

On a side note, how do know which triangles are part of the same n-gon? (Oh wait, I see. ab takes you to a neighboring triangle in the n-gon. Since (ab)^n=1, you can just cycle through.)

**Comment by ChristopherKing42** [comment]
Also, don't some members of the triangle group reverse orientation?

**Comment by ChristopherKing42** [comment]
Oh, and back on topic, addressing cells in the hyperbolic soccer ball isn't *too* tricky once you've done cell addressing in {7,3}. Just treat the heptagons in the soccer ball as the center of the heptagons in {7,3}, and treat the hexagons in the soccer ball as the vertices in of the heptagons in {7,3}.

Any [Truncation](https://en.wikipedia.org/wiki/Truncation_(geometry)) works similarly.

**Comment by dmishin** [comment]
You are absolutely right about (ab)^n, that's exactly how I do it in the code. You are right about the orientation too, I am only considering words of even length (i.e. orientation preserving).

Yes, cell addressing is really not _that_ much tricky. Existing code would work for a heptagon cells, and to get address of a hexagon, it should cycle over powers of (bc) instead of (ab). And address must be imbued with a boolean flag, telling whether it is address of a hexagon or a heptagon.

Maybe I'll try to implement it some day.

**Comment by ChristopherKing42** [comment]
If your worried about the faces being different, you could restrict it to [isohedral tessellations](https://en.wikipedia.org/wiki/Category:Isohedral_tilings). (For example, the dual of hyperbolic soccer ball is isohedral). Another interesting example is the [dual to the Alternated octagonal tiling](https://en.wikipedia.org/wiki/Alternated_octagonal_tiling#Dual_tiling), which is an isohedral hexagon tiling.
